scipy/fft/_pocketfft/scipy_features.h

```diff
@@ -0,0 +1,233 @@
+/** SciPy-only enhancements to pypocketfft
+
+ - Optimized full-sized fft for real input
+ - Normalization factor calculation
+
+*/
+
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <complex>
+
+#include "pocketfft.h"
+
+// Only instantiate long double transforms if they offer more precision
+using longfloat_t = typename std::conditional<
+  sizeof(long double) == sizeof(double), double, long double>::type;
+
+template<typename T> struct array_iterator
+  {
+  T * data, * cur;
+
+  pocketfft::stride_t strides;
+  pocketfft::shape_t shape, idx;
+
+  array_iterator(T * data_, pocketfft::stride_t strides_,
+                 pocketfft::shape_t shape_):
+    data(data_),
+    cur(data_),
+    strides(strides_),
+    shape(shape_),
+    idx(shape.size())
+    {
+    }
+
+  void reset()
+    {
+    std::fill(idx.begin(), idx.end(), 0);
+    cur = data;
+    }
+
+  /** Advance by 1, wrapping around into higher dimensions as necessary */
+  bool advance()
+    {
+    for (ptrdiff_t i = shape.size() - 1; i >= 0; --i)
+      {
+      ++idx[i];
+      cur += strides[i];
+
+      if (idx[i] < shape[i])
+        return true;
+
+      cur -= idx[i] * strides[i];
+      idx[i] = 0;
+      }
+    return false;
+    }
+
+  bool empty()
+    {
+    return std::any_of(shape.begin(), shape.end(),
+                       [](size_t len){ return len == 0; });
+    }
+
+  /** Take a slice along a single axis and reset index to (0,0,...,0) */
+  void slice(size_t dim, ptrdiff_t begin, ptrdiff_t end, ptrdiff_t step=1)
+    {
+    data += begin * strides[dim];
+    shape[dim] = end - begin;
+    strides[dim] *= step;
+    reset();
+    }
+  };
+
+enum class norm_t
+  {
+  none,  // No normalization
+  ortho, // Orthonormal transform
+  size   // Normalize by the fft size
+  };
+
+template<typename T> T norm_fct(norm_t norm, const pocketfft::shape_t & shape,
+                                const pocketfft::shape_t & axes)
+  {
+  if (norm == norm_t::none) return T(1);
+
+  T N(1);
```

Comment:
Sorry for noticing this so late ...

This snippet is actually problematic for single precision and large multi-D shapes, because the product of the axis lengths may no longer be representable exactly in `float`.
Also, doing the factor computation in the type of the transform introduces more rounding errors than strictly necessary.
How about this:
```
size_t N(1);
  for (auto a : axes)
    N *= shape[a];

  if (norm == norm_t::size) return T(1.L / N);
  if (norm == norm_t::ortho) return T(1.L / sqrt(1.L*N));
```

This may be slight overkill, but it doesn't really cost anything

Reply:
Thanks, fixed.